### PR TITLE
[Feature] Random erasing

### PR DESCRIPTION
**Pull request by Czm369**
image

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/datasets/transforms/transforms.py [hunk 1]
```diff
@@ -2878,3 +2878,157 @@ def __repr__(self):
...
+
+        self.n_patches = n_patches
+        self.ratio = ratio
+        self.squared = squared
+        self.bbox_erased_thr = bbox_erased_thr
+        self.img_border_value = img_border_value
+        self.mask_border_value = mask_border_value
+        self.seg_ignore_label = seg_ignore_label
+
+    def _get_patches(self, img_shape: Tuple[int, int]) -> List[list]:
```
Comment (RangiLyu) [inline mmdet/datasets/transforms/transforms.py:2956]:
```suggestion
    @cache_randomness
    def _get_patches(self, img_shape: Tuple[int, int]) -> List[list]:
```

File: mmdet/datasets/transforms/transforms.py [hunk 2]
```diff
@@ -2878,3 +2878,157 @@ def __repr__(self):
...
+            ``n_patches[1]``].
+        ratio (float or tuple[float, float]): The ratio of erased regions.
+            It can be ``float`` to use a fixed ratio or ``tuple[float, float]``
+            to randomly choose ratio from the interval.
+        squared (bool): Whether to erase square region. Defaults to True.
+        bbox_erased_thr (float): The threshold for the maximum area proportion
+            of the bbox to be erased. When the proportion of the area where the
+            bbox is erased is greater than the threshold, the bbox will be
+            removed. Defaults to 0.9.
+        img_border_value (int | float | tuple): The filled values for
```
Comment (RangiLyu) [inline mmdet/datasets/transforms/transforms.py:2919]:
```suggestion
        img_border_value (int or float or tuple): The filled values for
```

File: mmdet/datasets/transforms/transforms.py [hunk 3]
```diff
@@ -2878,3 +2878,158 @@ def __repr__(self):
...
+        img_border_value: Union[int, float, tuple] = 128,
+        mask_border_value: int = 0,
+        seg_ignore_label: int = 255,
+    ) -> None:
+        if isinstance(n_patches, tuple):
+            assert len(n_patches) == 2 and 0 <= n_patches[0] < n_patches[1]
+        else:
+            n_patches = (n_patches, n_patches)
+        if isinstance(ratio, tuple):
+            assert len(ratio) == 2 and 0 <= ratio[0] < ratio[1] < 1
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:2944]: Can't be equal to 1?

File: mmdet/datasets/transforms/transforms.py [hunk 4]
```diff
@@ -2878,3 +2878,158 @@ def __repr__(self):
...
+    def _transform_img(self, results: dict, patches: List[list]) -> None:
+        """Random erasing the image."""
+        for patch in patches:
+            px1, py1, px2, py2 = patch
+            results['img'][py1:py2, px1:px2, :] = self.img_border_value
+
+    def _transform_bboxes(self, results: dict, patches: List[list]) -> None:
+        """Random erasing the bboxes."""
+        bboxes = results['gt_bboxes']
+        left_top = np.maximum(bboxes[:, None, :2], patches[:, :2])
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:2987]: Please confirm whether there will be any problem with empty gt bbox, it is best to add unit test

File: mmdet/datasets/transforms/transforms.py [hunk 5]
```diff
@@ -2878,3 +2878,158 @@ def __repr__(self):
...
+        right_bottom = np.minimum(bboxes[:, None, 2:], patches[:, 2:])
+        wh = np.maximum(right_bottom - left_top, 0)
+        inter_areas = wh[:, :, 0] * wh[:, :, 1]
+        bbox_areas = (bboxes[:, 2] - bboxes[:, 0]) * (
+            bboxes[:, 3] - bboxes[:, 1])
+        bboxes_erased_ratio = inter_areas.sum(-1) / bbox_areas
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:2993]: inter_areas.sum(-1) / (bbox_areas+1e-7)

File: tests/test_datasets/test_transforms/test_transforms.py [hunk 6]
```diff
@@ -1116,3 +1116,36 @@ def test_repr(self):
...
+
+class TestRandomErasing(unittest.TestCase):
+
+    def setUp(self):
+        """Setup the model and optimizer which are used in every test method.
+
+        TestCase calls functions in this order: setUp() -> testMethod() ->
+        tearDown() -> cleanUp()
+        """
+        self.results = construct_toy_data(poly2mask=True)
```
Comment (hhaAndroid) [inline tests/test_datasets/test_transforms/test_transforms.py:1129]: empty bbox

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**